[CNSMR-3229] Add development process (#299)
Babylonpartners/ios-playbook: Cookbook/Technical-Documents/Development-Process.md

# Development Process

Babylon follows the [SSDLC](https://engineering.ops.babylontech.co.uk/docs/standards-ssdlc) when developing our software. To make the process easier, we use a few services to link code changes to tickets.

There are 3 services we use daily:

- Jira: this is where the development begins. Once a ticket has all the necessary details, dev can pick it up and implement it. <!-- TODO: [CNSMR-3230] link to Jira article -->
- GitHub: used for hosting our project
- CircleCI: our continuous integration service to run tests when integrating PRs <!-- TODO: [CNSMR-xxxx] link to CircleCI article -->

> ## Jira 🎫

Reviewer: ```suggestion
## 1. Get a Jira ticket 🎫
```

- Everything we do must have a ticket

> - The first step is to assign the ticket to the engineer. Make sure it has all details and designs attached (if applicable)

Reviewer: ```suggestion
- The ticket must be assigned to the engineer who is going to work on it. The engineer should make sure it has all details and designs attached (if applicable).
```

> - The ticket is moved to `In progress` (this might be automatically done by the bot when branch published, if configured)

Reviewer: ```suggestion
- The ticket will be moved to `In progress` once development is ready to start (this might be automatically done by the bot when branch published, if configured).
```

> ## Get started with development 🛠

Reviewer: ```suggestion
## 2. Get started with development 🛠
```

> - Create a branch where this work will live. Please follow this format: `[NAME]/[TICKET]-[DESCRIPTION (optional)]` (E.g. `john/TCKT-123-fix-bug`)

Reviewer: ```suggestion
- Create a branch where this work will live. Please follow this format: `<author's_name>/<ticket-id>-<small-description (optional)>` (E.g. `john/TCKT-123-fix-bug`)
```

- For new features, add a [local feature switch](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/FeatureSwitches.md) to avoid getting WIP code into the release. Add the name of this flag (and additional instructions) to the ticket to help QA test it

> - Implement all features, changes to satisfy the acceptance criteria defined in the ticket

Reviewer 2: Would a link to some of our architecture patterns like Unidirectional flow, Reactive Swift, Bento, ReactiveFeedback etc under this section also be useful?  I think "Get Started with development"  could be split into "Git setup and feature flagging" and then "Development"(which includes links to best practices, architecture documentation, meeting A/C etc).

Reviewer 3: Wasn't sure how detailed this article should be. I think if someone's looking for code guides, they can find it in the Table of Contents, otherwise when we create or edit articles, it'd need to be updated in both places (and there's a good chance it won't happen). I'd rather keep this simple and generic

Reviewer: I agree with you Dan. We can mention:
> the implementation should be based in our current [Architecture](https://github.com/babylonhealth/babylon-ios/blob/develop/Documentation/README.md#architecture-)

Reviewer 3: Just going to wait until #301 is merged so I can link to the section :)

Reviewer: yeah makes sense :)

> - If it introduces a new component, add it to the `GalleryApp` and make sure it's included in the Design System

Reviewer: Can we add a link to the `GalleryApp`?

Reviewer 3: In App Center? Or do we have an article about it?

Reviewer: App Center as there is not article for it ¯\_(ツ)_/¯

Reviewer 3: Fixed in 15b472e

> ## Testing 🧪 <!-- TODO: [CNSMR-3195] link to testing article -->

Reviewer: ```suggestion
## 3. Testing 🧪 <!-- TODO: [CNSMR-3195] link to testing article -->
```

> - Make sure:

Reviewer: ```suggestion
- Our features should include the following type of tests:
```

>   - New components, screens have snapshot tests

Reviewer: ```suggestion
  - Snapshot test should be added to new components and screens
```

>   - Logic, ViewModels, SDK changes are unit tested

Reviewer: ```suggestion
  - Logic, ViewModels, SDK changes should be unit tested
```

>   - Automation tests have been updated or added if the feature will be in the next release

Reviewer: ```suggestion
  - Automation tests should be updated or added.
```

> - Build `GalleryApp` and `Babylon` before merging to avoid checks failing

Reviewer: ```suggestion
- Please test `GalleryApp` and `Babylon` by building and running before merging to avoid checks failing.
```

> ## Creating the PR 📝

Reviewer: ```suggestion
## 4. Creating the PR 📝
```

- Fill in the PR template and set the title using the following format: `[TCKT-123] Short summary of work`
- [Add labels](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/LabelsInPRs.md)
- Assign the PR to the engineer(s)
- Move the ticket to `Peer review` on Jira (or it'll be automatically moved if configured)
- If you'd like feedback before the work on the ticket is done, open a draft PR so your peers can leave comments
- To run checks before merging, use the comment `@ios-bot-babylon test_pr` ([more info](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/SlackCIIntegration.md))

_Note: Sometimes Pull Assigner is not triggered when a new PR is opened. In that case, re-request `ios-pullassigner` to add reviewers from the team._

> ## Merging PR 🚦

Reviewer: ```suggestion
## 5. Merging PR 🚦
```

- Once a PR has 2 approvals and no outstanding comments/changes requested, the `Merge` label can be added to begin the merging process
- Our [bot](https://github.com/babylonhealth/Wall-E) will add it to the merge queue and integrate it into the rest of the project

> - Once it reaches the front, it's updated with the target branch and CircleCI runs all necessary tests to make sure nothing's failing. _Note: there's no need to trigger checks manually, if there are other PRs in the queue._ <!-- TODO: [CNSMR-3231] link to Wall-E article -->

Reviewer: ```suggestion
- Once it reaches the front of the queue, it's updated with the target branch and CircleCI runs all necessary tests to make sure nothing's failing. _Note: there's no need to trigger checks manually, if there are other PRs in the queue._ <!-- TODO: [CNSMR-3231] link to Wall-E article -->
```

- If the required checks are passing, the PR is merged and the branch deleted.
- The ticket is moved to `Awaiting build` on Jira by our [bot](https://github.com/babylonhealth/Stevenson). It will be updated to `Ready for QA` when the next App Center build is created. (if configured, otherwise this should be done manually)

> 

Reviewer: ```suggestion
- If the required checks fail you will have to analyse what is the reason for the failure, fix it and add the `Merge` label again.
```

Reviewer 3: Checking the "required checks" part, there was a recent change in Wall-E that might have changed this

Reviewer 3: All checks are needed, not just required ones
```suggestion
- If the checks fail, you will have to analyse what is the reason for the failure, fix it, and add the `Merge` label again
```

> ## QA 🧑‍💻

Reviewer: ```suggestion
## 6. QA 🧑‍💻
```

- Once the work is in the App Center builds, QA can test it to make sure everything works correctly
- It's important that the engineer assigned to the ticket doesn't QA their own work
- When QA is done with testing, the ticket can be moved to `Done` on Jira

> ## Release 🚢

Reviewer: ```suggestion
## 7. Release 🚢
```

> - Make sure all local feature flags have been removed **before** the release branch is cut

Reviewer: ```suggestion
- When releasing a feature that was under a local feature flag, make sure it is removed **before** the release branch is cut
```

# Development Process

At Babylon security is something we have always in mind when developing software and that is why we follow the SSDLC (Secure Software Development Life Cycle) framework.
Alongside this framework, we also use a few services to help us keep track of what we are building and make sure we deliver a secure and stable product.

There are 3 services we use daily:

- Jira: is where the development process begins. Once a ticket has all the necessary details, dev can pick it up and implement it <!-- TODO: [CNSMR-3230] link to Jira article -->
- GitHub: used for hosting our project and version control using Git
- CircleCI: our continuous integration service to run tests when integrating PRs <!-- TODO: [CNSMR-xxxx] link to CircleCI article -->

## 1. Get a Jira ticket 🎫

- Every development task that results in a PR must have a Jira ticket
- The ticket must be assigned to the engineer who is going to work on it. The engineer should make sure it has all details and designs attached (if applicable).

> - The ticket will be moved to `In progress` once development is ready to start (this might be automatically done by the bot when branch published, if configured)

Reviewer 4: ```suggestion
- The ticket will be moved to `In progress` once development is ready to start (this might be automatically done by the bot when branch is published, if configured)
```
